WebGLRenderer.renderOverride - make special purpose scene renders one line

bhouston opened the pull request:
Prior to this PR it was necessary to setup the clear color, clear alpha and autoClear variables as well as setting and clearing the overrideMaterial when rendering a special purpose pass.  It looked like this:

```
scene.overrideMaterial = depthMaterial;
renderer.setClearColor( 0xffffff );
renderer.setClearAlpha( 0 );
renderer.render( scene, camera, renderTarget, true );       
scene.overrideMaterial = null;      
```

(BTW the render the clear color and alpha needed to be set is that for passes like depth one needs to have it cleared to 1,1,1, which is a non-standard clear color.  If you do not set the clear color you are dependent on whatever the clear color was, which may lead to bad results.)

With this PR the above five lines becomes, and you never need to remember to clear the `scene.overrideMaterial` variable:

```
renderer.renderOverride( depthMaterial, scene, camera, renderTarget, 0xffffff, 0 );     
```

This pairs really nicely with https://github.com/mrdoob/three.js/pull/8625.

bhouston committed:
add WebGLRenderer.renderPass

bhouston committed:
clean up clearColor checking in renderPass

bhouston committed:
fix bug in ShaderPass with adoption of WebGLRenderer.renderPass

bhouston committed:
fix bug in renderPass.

bhouston committed:
WebGLRenderer.renderOverride - make it a one liner to use override materials.

# Conflicts:
#	examples/js/postprocessing/BokehPass.js
#	src/renderers/WebGLRenderer.js

bhouston committed:
improve renderOverride (fix bugs)

bhouston committed:
shorten up code.

bhouston committed:
shorten up code.

tschw commented:
Would you always use a `depthMaterial`? (This might be about naming the beast, I guess).
Would you always want to clear? Or might the user want to blend on the current target as well?

bhouston commented:
> Would you always use a depthMaterial? (This might be about naming the beast, I guess).

I use MeshNormalMaterial, MeshDepthMaterial are the two most common for sure.  But if you look at the changed files you can see a variety of materials used:

https://github.com/mrdoob/three.js/pull/8626/files

> Would you always want to clear? Or might the user want to blend on the current target as well?

The current code supports both workflows for full backwards compatibility (clear or no clear.)  I am unsure if blending on the current target is something that is done often with overrideMaterials - I do not have experience with that use case personally.

tschw commented:
Was thinking your suggestion might also be useful for postprocessing, when reordering the parameters and making some optional, where no scene means UDC plane...

bhouston commented:
You need this PR? https://github.com/mrdoob/three.js/pull/8625

tschw commented:
I see :+1:

bhouston committed:
do not clear unless explicitly requested to clear.

bhouston committed:
only clear if explicitly requested to clear.

bhouston commented:
@tschw One question though is that there is a specific background color to use when doing a normal pass, it is new THREE.Color( 0.5, 0.5, 1.0 ) -- which is a normal facing the viewer --, and for depth the default color and alpha needs 0xffffff, and 1.0 -- which is the far clipping plane.  Basically anything else is wrong.

I wonder how I should try to standardize these.  MeshDepthMaterial.ClearColor, MeshDepthMaterial.ClearAlpha as constants maybe?  Or could make them members variables of the Depth and Normal materials and when a clearAlpha, clearColor is encountered by the renderOverride function on an overrideMaterial it uses them as the clear color and clear alpha?

bhouston committed:
add default clearColor, clearAlpha support for override materials to make depth and normal passes easier to get right.

bhouston commented:
I've just added the material.clearColor/clearAlpha behavior -- it is fairly elegant.

bhouston commented:
On top of these two recent PRs, renderOverride + renderPass, I've built a fully arbitrary kernel size Gaussian blur filter:

![image](https://cloud.githubusercontent.com/assets/588541/14506306/14f36884-018b-11e6-9572-dee3762bbe65.png)

![image](https://cloud.githubusercontent.com/assets/588541/14506278/f51d3710-018a-11e6-86f5-f3e26545c14f.png)

![image](https://cloud.githubusercontent.com/assets/588541/14506300/0d8c3a12-018b-11e6-9c66-393394748e58.png)

![image](https://cloud.githubusercontent.com/assets/588541/14506374/734997aa-018b-11e6-9e4d-38968f16a3ce.png)

tschw commented:
> On top of these two recent PRs, renderOverride + renderPass, I've built a fully arbitrary kernel size Gaussian blur filter

Man, you're really hyperactive these days :).

60fps at scale (!). Optimized? At least using an X/Y separated kernel, I guess...

bhouston committed:
fix typo.

bhouston committed:
WebGLRenderer.post* -> .pass* for renderPass consistent naming.

bhouston commented:
@mrdoob, I'm back for a bit.  I do suggest trying to handle these PRs separately so I can get one or a few merged.  That mega SAO one is a bit overwhelming.

Do you want this rendererOverride on WebGLRenderer if I separate out the clear color/alpha stuff?

Do you want the renderPass stuff as well in this other PR if I make a similar change?  e.g. https://github.com/mrdoob/three.js/pull/8625

mrdoob commented:
> Do you want this rendererOverride on WebGLRenderer if I separate out the clear color/alpha stuff?

Yes!

> Do you want the renderPass stuff as well in this other PR if I make a similar change? e.g. #8625

Yes yes!

We could even merge both into a single name...  `.renderMaterial( material, scene, camera )`. If `scene` isn't sent then we render a internal scene/camera which is a quad.

bhouston committed:
Merge branch 'renderPass' of git@github.com:bhouston/three.js into renderOverride

# Conflicts:
#	src/renderers/WebGLRenderer.js

bhouston committed:
adopt WebGLRendererUtils.

bhouston committed:
update examples to use WebGLRendererUtils properly.

bhouston committed:
tabify.

bhouston commented:
@mrdoob, I've tried an experiment based on your feedback.

I've removed renderPass/renderOverride from WebGLRenderer and moved them into a new class /examples/js/WebGLRendererUtils.js:

https://github.com/bhouston/three.js/blob/renderOverride/examples/js/WebGLRendererUtils.js

Neither renderPass or renderOVerride are used by /src anyhow.  They are just used for simplifying post processing.  This allows me to keep the easy-to-use interface that includes the clear colors, which makes post processing code very short to write -- which is my main goal, short expressive code.  :)

I have also merged renderPass into this branch.  Let me know if you like this formulation?

mrdoob commented on examples/js/Ocean.js:
```diff
@@ -253,16 +250,14 @@ THREE.Ocean.prototype.generateSeedPhaseTexture = function() {
 	this.materialInitialSpectrum.uniforms.u_wind.value.set( this.windX, this.windY );
 	this.materialInitialSpectrum.uniforms.u_size.value = this.size;
-	this.renderer.render( this.scene, this.oceanCamera, this.initialSpectrumFramebuffer, true );
+	this.THREE.WebGLRendererUtils.renderOverride( renderer, this.materialInitialSpectrum, this.scene, this.oceanCamera, this.initialSpectrumFramebuffer, true );
```
`this.THREE`?

bhouston replied:
Fixed that.  So you are okay with the refactor our of WebGLRenderer?  Sweet.

I've renamed WebGLRendererUtils to be EffectRenderer and moved it into postprocessing -- this makes sense to me.  It is just an easy way to do post effects anyhow without resorting to Pass-based system.

bhouston committed:
rename/move /WebGLRenderUtils.js -> postprocessing/EffectRenderer.js

bhouston committed:
cleanup.

bhouston committed:
remove unused variables.

bhouston committed:
more clean-up.

mrdoob commented:
Looking good so far!

mrdoob commented on src/materials/MeshDepthMaterial.js:
```diff
@@ -33,6 +33,10 @@ THREE.MeshDepthMaterial = function ( parameters ) {
 
+	// far clipping plane in both RGBA and Basic encoding
+	this.clearColor = new THREE.Color( 1.0, 1.0, 1.0 );
+	this.clearAlpha = 1.0;
```
When are these needed?

bhouston replied:
They are automatically used by EffectRenderer if you do not specify overrides when using renderPass, renderOverride:

https://github.com/mrdoob/three.js/pull/8626/files/53b5fb0548bd0a82494d3bd7f2e68ec24b15b1f7#diff-2bb4aa115364b94510e8f396157285c7R37

Most people do not know that you have to clear depth so that it defaults to a far distance where polygons are not written.  The same with the normal material's clear color/clearAlpha: 

https://github.com/mrdoob/three.js/pull/8626/files/53b5fb0548bd0a82494d3bd7f2e68ec24b15b1f7#diff-717b15a206ca91fed8cb9df26d53be71R23

We do not need to have them in the materials, but it is really useful to have these in the materials themselves as they are really required for that material to work properly.  If you clear depth with black you are saying everything is close, which is generally wrong.  If you clear normal with black or white, you are clearing it with what are invalid normals, which will cause issues when you try to make sense of them in a post effect.

bhouston commented:
Superceded by simplier PR: https://github.com/mrdoob/three.js/pull/8913
